go/vt/vtadmin/cluster/cluster.go

```diff
@@ -373,50 +374,269 @@ func (c *Cluster) getTablets(ctx context.Context) ([]*vtadminpb.Tablet, error) {
 	}
 
-	if schema == nil || schema.Schema == nil {
-		return nil, nil
+	var tabletsToQuery []*vtadminpb.Tablet
+
+	if opts.SizeOpts.AggregateSizes {
+		span, ctx := trace.NewSpan(ctx, "Cluster.FindAllShardsInKeyspace")
+
+		AnnotateSpan(c, span)
+		span.Annotate("keyspace", keyspace)
+
+		resp, err := c.Vtctld.FindAllShardsInKeyspace(ctx, &vtctldatapb.FindAllShardsInKeyspaceRequest{
+			Keyspace: keyspace,
+		})
+
+		span.Finish()
+
+		if err != nil {
+			return nil, fmt.Errorf("FindAllShardsInKeyspace(cluster = %s, keyspace = %s) failed: %w", c.ID, keyspace, err)
+		}
+
+		for _, shard := range resp.Shards {
+			if !shard.Shard.IsMasterServing {
+				if !opts.SizeOpts.IncludeNonServingShards {
+					log.Infof("%s/%s is not serving; ignoring because IncludeNonServingShards=false", keyspace, shard.Name)
+					continue
+				}
+			}
+
+			shardTablets := vtadminproto.FilterTablets(func(tablet *vtadminpb.Tablet) bool {
+				return tablet.Tablet.Shard == shard.Name && tablet.State == vtadminpb.Tablet_SERVING
+			}, opts.Tablets, len(opts.Tablets))
+
+			if len(shardTablets) == 0 {
+				return nil, fmt.Errorf("%w for shard %s/%s", errors.ErrNoServingTablet, shard.Keyspace, shard.Name)
+			}
+
+			randomServingTablet := shardTablets[rand.Intn(len(shardTablets))]
+			tabletsToQuery = append(tabletsToQuery, randomServingTablet)
+		}
+	} else {
+		keyspaceTablets := vtadminproto.FilterTablets(func(tablet *vtadminpb.Tablet) bool {
+			return tablet.Tablet.Keyspace == keyspace && tablet.State == vtadminpb.Tablet_SERVING
+		}, opts.Tablets, len(opts.Tablets))
+
+		if len(keyspaceTablets) == 0 {
+			// consider how to include info about the tablets we looked at, but
+			// that's also potentially a very long list .... maybe we should
+			// just log it (yes, do that).
+			return nil, fmt.Errorf("%w for keyspace %s", errors.ErrNoServingTablet, keyspace)
+		}
+
+		randomServingTablet := keyspaceTablets[rand.Intn(len(keyspaceTablets))]
+		tabletsToQuery = append(tabletsToQuery, randomServingTablet)
 	}
 
-	return &vtadminpb.Schema{
-		Cluster:          c.ToProto(),
-		Keyspace:         tablet.Tablet.Keyspace,
-		TableDefinitions: schema.Schema.TableDefinitions,
-	}, nil
+	var (
+		m      sync.Mutex
+		wg     sync.WaitGroup
+		rec    concurrency.AllErrorRecorder
+		schema = &vtadminpb.Schema{
+			Cluster:    c.ToProto(),
+			Keyspace:   keyspace,
+			TableSizes: map[string]*vtadminpb.Schema_TableSize{},
+		}
+		// Instead of starting at false, we start with whatever the base request
+		// specified. If we have exactly one tablet to query (i.e. we're not
+		// doing multi-shard aggregation), it's possible the request was to
+		// literally just get the table sizes; we shouldn't assume. If we have
+		// more than one tablet to query, then we are doing size aggregation,
+		// and we'll flip this to true after spawning the first GetSchema rpc.
+		sizesOnly = opts.BaseRequest.TableSizesOnly
+	)
+
+	for _, tablet := range tabletsToQuery {
```

Comment:
Here similar to my other comment about throttling. I'm not too concern, because I think this should scale pretty well as it's rpc to different hosts. But it does makes a bit nervous that for large keyspaces we can be making too many of this in parallel. 

Wondering if putting a guardrail where we say: we only do X amount of parallelism will be good to have in place early on.

Reply:
Yeah, I think when we do the "bound topo RPCs to max concurrency" there should be two limits: tablet RPC max concurrency, and topo RPC max concurrency (maybe others). Then each of those gets configured (per cluster, so larger clusters could in theory have higher limits) and we spawn a waitgroup/pool/etc with that value as the capacity.